etcdserver/corrupt.go

```diff
@@ -19,11 +19,65 @@ import (
 	"time"
 
 	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes"
 	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
 	"github.com/coreos/etcd/mvcc"
 	"github.com/coreos/etcd/pkg/types"
 )
 
+// checkInitialHashKV compares initial hash values with its peers
+// before serving any client traffic.
+func (s *EtcdServer) checkInitialHashKV() {
+	if !s.Cfg.InitialCorruptCheck {
+		return
+	}
+
+	// wait until server is ready by publishing attributes
+	// to "/0/members/*", so that *membership.RaftCluster
+	// is up-to-date with its peers, on restart
+	select {
+	case <-s.readych:
+	case <-s.stopping:
+		return
+	}
+
+	plog.Infof("%s starting initial corruption check...", s.ID())
+	h, rev, crev, err := s.kv.HashByRev(0)
+	if err != nil {
+		plog.Fatalf("%s failed to fetch hash (%v)", s.ID(), err)
+	}
+
+	peers := s.getPeerHashKVs(rev)
+	mismatch := 0
+	for _, p := range peers {
+		if p.resp == nil {
```

Comment:
probably handle p.resp != nil case first. it is simpler and is the most common case. it improves readability.